File: server/src/main/kotlin/schema/Snapshot.kt
//
//  Snapshot.kt
//  incicord
//
//  Created by d-exclaimation on 4:41 PM.
//

package schema

/**
 * ### Event Snapshot
 *
 * Message JSON sent through websocket
 */
data class Snapshot<T>(
    val operation: String,
    val payload: T
) {
    companion object {
        /**
         * Creation Snapshot constructor
         */
        fun <T> creation(payload: T): Snapshot<T> = Snapshot(
            operation = "creation",
            payload = payload
        )

        /**
         * Mutation Snapshot constructor
         */
        fun <T> mutation(payload: T): Snapshot<T> = Snapshot(
            operation = "mutation",
            payload = payload
        )

        /**
         * Deletion Snapshot constructor
         */
        fun <T> deletion(payload: T): Snapshot<T> = Snapshot(
            operation = "deletion",
            payload = payload
        )
    }
}

File: web/src/snapshot/Snapshot.ts
//
//  Snapshot.ts
//  web
//
//  Created by d-exclaimation on 22:46.
//

export type Snapshot<
  TCreation extends object,
  TMutation extends object,
  TDeletion extends object
> =
  | { operation: "creation"; payload: TCreation }
  | { operation: "deletion"; payload: TDeletion }
  | { operation: "mutation"; payload: TMutation };

export function Snapshot<
  TCreation extends object,
  TMutation extends object,
  TDeletion extends object
>(json: string): Snapshot<TCreation, TMutation, TDeletion> {
  return JSON.parse(json);
}

export function handleEvent<
  TCreation extends object,
  TMutation extends object,
  TDeletion extends object
>(
  snap: Snapshot<TCreation, TMutation, TDeletion>,
  handler: SnapshotHandlers<TCreation, TMutation, TDeletion> = defaultHandler
): void {
  switch (snap.operation) {
    case "creation":
      return handler.creation(snap.payload);
    case "mutation":
      return handler.mutation(snap.payload);
    case "deletion":
      return handler.deletion(snap.payload);
    default:
      throw new Error("Wtf");
  }
}

export type SnapshotHandlers<
  TCreation extends object,
  TMutation extends object,
  TDeletion extends object
> = {
  creation: (payload: TCreation) => void;
  mutation: (payload: TMutation) => void;
  deletion: (payload: TDeletion) => void;
};

const defaultHandler: SnapshotHandlers<any, any, any> = {
  creation: () => {},
  mutation: () => {},
  deletion: () => {},
};
